logging/impl.go

```diff
@@ -41,6 +59,32 @@ type (
 	}
 )
 
+// String stringifies a `LogEntry`. Should be used to emplace a log entry in
+// `recentMessageEntries`, i.e. `LogEntry`s that `String`ify identically should
+// be treated as identical with respect to noisiness and deduplication. Should
+// not be used for actual writing of the entry to an appender.
+func (le *LogEntry) String() string {
```

Comment:
IIUC, this is for computing the key "string" for `recentMessageEntries`. I don't think this needs any of the special formatting?

Two remarks:
* If we're going to use this method for creating a special/unique key, let's not call it a `String` method. For all of the reasons you wrote above -- the consumer of this method requires some crucial properties (great job describing them!). A method named `String` generally is assumed to not be bound by such strict requirements.
* But more preferably -- I'm a little concerned about the performance impact of this. More about the number of small allocations rather than CPU time. I'd recommend looking into the [hashing functionality Go provides](https://pkg.go.dev/hash/maphash#Hash) to create a unique identifier for a `LogEntry`.
  * I don't need to see a benchmark, but I think* Go makes it easy to do memory allocation benchmarks with the `-benchmem` flag. It that sounds like fun.